src/infra/core/src/dependency_graph_service_inmem.rs

```diff
@@ -193,6 +194,102 @@ impl DependencyGraphService for DependencyGraphServiceInMemory {
             .await
     }
 
+    async fn get_all_upstream_dependencies(
+        &self,
+        dataset_ids: Vec<DatasetID>,
+    ) -> Result<DatasetIDStream, GetUpstreamDependenciesError> {
+        self.ensure_datasets_initially_scanned()
+            .await
+            .int_err()
+            .map_err(GetUpstreamDependenciesError::Internal)
+            .unwrap();
+
+        let state = self.state.lock().await;
+
+        let nodes_to_search = dataset_ids
+            .iter()
+            .map(|dataset_id| {
+                state
+                    .get_dataset_node(dataset_id)
+                    .map_err(GetUpstreamDependenciesError::DatasetNotFound)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        let mut node_indexes_result = HashSet::new();
+
+        for node_index in &nodes_to_search {
+            let mut bfs = Bfs::new(&state.datasets_graph, *node_index);
```

Comment:
I think that `Bfs` vs `Dfs` only defines the order of visiting the graph nodes (breadth-first vs depth-first). It does not define directions of traversal like you assume here.

I don't think we care about the exact traversal algorithm here, all we want is a set, so pick any.
However, in order to invert the direction of traversal, I think you'll need something like [Reversed ](https://docs.rs/petgraph/latest/petgraph/visit/struct.Reversed.html), but you have to experiment.